bip-schnorr.mediawiki

```diff
@@ -36,6 +36,7 @@ made:
 
 * '''Signature encoding''': Instead of using [https://en.wikipedia.org/wiki/X.690#DER_encoding DER]-encoding for signatures (which are variable size, and up to 72 bytes), we can use a simple fixed 64-byte format.
 * '''Public key encoding''': Instead of using ''compressed'' 33-byte encodings of elliptic curve points which are common in Bitcoin today, public keys in this proposal are encoded as 32 bytes.
+* '''Consistent validity''': Traditional properties demanded from a digital signature scheme include correctness (an honest signature will be valid to all verifiers) and unforgeability (someone without the private key cannot construct a signature that is valid to any verifier). For expected use in a consensus system, we need an additional property: consistent validity (nobody - with or without private key - can construct a signature that is valid to some but not all verifiers), to prevent malicious signers from splitting the network. The fact that parsing of ECDSA signature is not exactly specified made this property [https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2015-July/009697.html broken] in the past, needing [https://github.com/bitcoin/bips/blob/master/bip-0146.mediawiki BIP66] to address it. In this document we aim to meet this property by design.
```

Comment:
nit: s/honest signature/honestly created signature/ ?

Does Consistent Validatity mean more than just saying that the bip-schnorr algorithms take bytes as input and output a boolean or bytes? If not then we could just say something like *Complete Spec* instead of the more vague *Consistent validity*.

Reply:
I agree. 

You phrase this as a security property with an attacker in mind. While I agree with the goal, I don't think we need to mention this. On a technical level consistent validity simply boils down to making sure that every verifier runs the same algorithm. So it's really only "completely specified" or "exact specification" or "specified on bytes".

Reply:
That's fair. I had slightly more in mind than exactly specified, though - it also either needs to be deterministic or the nondeterminism may not make it easy for someone with a private key to construct a signature that has more-than-negligable-probability for both being valid and invalid. Given that the batch verification algorithm is actually nondeterministic, I don't think it's as simple as stating it's exactly specified.

Reply:
Hm okay, the non-deterministic batch verification is something one needs to consider... That's somewhat weird to be honest. Assuming it is really randomized, I would not change much of the big picture because then two verifiers differ only with negl. probability. The same should hold in the ROM when deriving the seed as we currently specify it in the BIP. (The attacker can now try many hashes locally but this gives it only a polynomial advantage.)  

But this raises a valid point: If we think that verification should be consistent, shouldn't we simply prescribe the exact algorithm even for batch verification? (We can't force people to use our implementation of course, but we can't do that anyway.)

Reply:
" correctness (an honest signature will be valid to all verifiers)"  < should that instead say "signature created with a specified algorithm.  That's also what "honest" means, but I think in this case it is unneeded domain specific language.

I also like the term "completely specified"

> consistent validity simply boils down to making sure that every verifier runs the exact same algorithm

Not exactly the same, but functionally equivalent.

Reply:
I like "completely specified", but still think it needs to be more specific (no pun intended) about how to deal with non-deterministic algorithms. I'll write something.

> But this raises a valid point: If we think that verification should be consistent, shouldn't we simply prescribe the exact algorithm even for batch verification?

That won't achieve non-determinism, as the batches themselves remain a verifier-side choice (if they have pre-verified transactions in the mempool, those don't need to be re-validated again in a block, for example, and in IBD a node may well want to choose to batch signatures from 1000 blocks into one). I also don't think it's necessary.

Reply:
The batch verification ends up having a 1:2^[128,256] probability of falsely accepting an invalid batch, and it'll never falsely reject.  This is probably a lot better than the underlying hardware's reliability.  The actual property we want is that the probability two conforming validators can disagree even in the presence of entirely adversarial inputs is negligible.

Reply:
I've rewritten things, have another look.